README.md
# react-native-material-backdrop-modal
Material Design "Backdrop" component for Android and iOS.

- [Setup](#setup)
- [Usage](#usage)
- [Props](#props)

## Setup
1. Install:
    - Using [npm](https://www.npmjs.com/#getting-started): `npm install react-native-material-backdrop-modal --save`
    - Using [Yarn](https://yarnpkg.com/): `yarn add react-native-material-backdrop-modal`

2. Import it in your JS:
    ```js
    import Backdrop from 'react-native-material-backdrop-modal';
    ```

## Usage
A backdrop appears behind all other surfaces in an app, displaying contextual and actionable content.

![Backdrop](./example/backdrop.png)

See Google's [Material Design guidelines](http://material.io/components/backdrop/) for more info about Backdrops and when to use them.

Check out the example app:

```bash
git clone https://github.com/brunohkbx/react-native-material-backdrop-modal
cd react-native-material-backdrop-modal/example
yarn
yarn start
```

![Example](./example/example.gif)

## Props
| prop                      | default                  | type          | description                                                                                                                                                     |
| ------------------------- | ------------------------ | ------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| children                  |                  | node        | Content of `Backdrop     `                                                                                                                                       |
| focused            | true                     | boolean        | Specifies whether `Backdrop` should be focused                                                                                                                                        |
| onFocus           |                      | Function        | The Handler that's emitted every time the user conceals the back layer.                                                                                                                          |
| title         |  | string        | The Subheader title                                                                                                                             |
| icon | | element | Icon to be used on the Subheader
| preset | easeInEaseOut | object | Animation [presets](https://reactnative.dev/docs/layoutanimation#presets) to pass into configureNext
| titleStyle | {} | object | Style for the subheader title
| backdropStyle |{} | object | Style for the backdrop


CHANGELOG.md
## [1.6.4](https://github.com/brunohkbx/react-native-material-backdrop-modal/compare/v1.6.3...v1.6.4) (2022-11-28)


### Bug Fixes

* PropTypes for styles;accept arrays ([f82977f](https://github.com/brunohkbx/react-native-material-backdrop-modal/commit/f82977f0ce123dc6ddb72f34e3913e7b95f609f3))

## [1.6.3](https://github.com/brunohkbx/react-native-material-backdrop-modal/compare/v1.6.2...v1.6.3) (2021-05-15)


### Bug Fixes

* remove wrong padding from children container ([1a07a88](https://github.com/brunohkbx/react-native-material-backdrop-modal/commit/1a07a88c70bcc94425175b995cfd217ac65cff6c))

## [1.6.2](https://github.com/brunohkbx/react-native-material-backdrop-modal/compare/v1.6.1...v1.6.2) (2021-05-15)


### Bug Fixes

* TalkBack/VoiceOver focusing on elements below backdrop ([bd427d6](https://github.com/brunohkbx/react-native-material-backdrop-modal/commit/bd427d650d0da8e6883bd18fef873557c631920f))


__tests__/Backdrop.test.js
import React from 'react';
import { View, Text } from 'react-native';
import { fireEvent, render } from 'react-native-testing-library';
import Backdrop from '../src/Backdrop/Backdrop';

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
jest.mock('react-native/Libraries/LayoutAnimation/LayoutAnimation');

describe('Backdrop', () => {
  const setup = (propOverrides = {}) => {
    const defaultProps = {
      onFocus: jest.fn(),
      title: 'Movies',
      ...propOverrides,
    };

    const element = (
      <Backdrop {...defaultProps}>
        <View>
          <Text>Movie 1</Text>
        </View>
      </Backdrop>
    );

    const wrapper = render(element);

    return { element, wrapper };
  };

  describe('When updating from focused to unfocused', () => {
    it("changes the backdrop's flex property", () => {
      const {
        wrapper: { getByTestId, update },
      } = setup({ focused: true });

      const { element: updatedElement } = setup({ focused: false });
      update(updatedElement);

      expect(getByTestId('backdrop').props.style).toContainEqual({
        flex: 0,
      });
    });
  });

  describe('When updating from unfocused to focused', () => {
    it("changes the backdrop's flex property", () => {
      const {
        wrapper: { getByTestId, update },
      } = setup({ focused: false });

      const { element: updatedElement } = setup({ focused: true });
      update(updatedElement);

      expect(getByTestId('backdrop').props.style).toContainEqual({
        flex: 1,
      });
    });
  });

  describe('onFocus', () => {
    it('calls onFocus when pressing on Subheader', () => {
      const mockOnFocus = jest.fn();
      const {
        wrapper: { getByTestId },
      } = setup({ focused: false, onFocus: mockOnFocus });

      fireEvent.press(getByTestId('subheader'));

      expect(mockOnFocus).toHaveBeenCalled();
    });

    it("calls onFocus when pressing on Subheader's icon", () => {
      const mockOnFocus = jest.fn();
      const {
        wrapper: { getByTestId },
      } = setup({ focused: false, onFocus: mockOnFocus, icon: <View /> });

      fireEvent.press(getByTestId('icon'));

      expect(mockOnFocus).toHaveBeenCalled();
    });
  });
});


src/Backdrop/__tests__/Backdrop.test.js
import React from 'react';
import { View, Text } from 'react-native';
import { fireEvent, render } from 'react-native-testing-library';
import Backdrop from '../Backdrop';

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
jest.mock('react-native/Libraries/LayoutAnimation/LayoutAnimation');

describe('Backdrop', () => {
  const setup = (propOverrides = {}) => {
    const defaultProps = {
      onFocus: jest.fn(),
      title: 'Movies',
      ...propOverrides,
    };

    const element = (
      <Backdrop {...defaultProps}>
        <View>
          <Text>Movie 1</Text>
        </View>
      </Backdrop>
    );

    const wrapper = render(element);

    return { element, wrapper };
  };

  describe('When updating from focused to unfocused', () => {
    it("changes the backdrop's flex property", () => {
      const {
        wrapper: { getByTestId, update },
      } = setup({ focused: true });

      const { element: updatedElement } = setup({ focused: false });
      update(updatedElement);

      expect(getByTestId('backdrop').props.style).toContainEqual({
        flex: 0,
      });
    });
  });

  describe('When updating from unfocused to focused', () => {
    it("changes the backdrop's flex property", () => {
      const {
        wrapper: { getByTestId, update },
      } = setup({ focused: false });

      const { element: updatedElement } = setup({ focused: true });
      update(updatedElement);

      expect(getByTestId('backdrop').props.style).toContainEqual({
        flex: 1,
      });
    });
  });

  describe('onFocus', () => {
    it('calls onFocus when pressing on Subheader', () => {
      const mockOnFocus = jest.fn();
      const {
        wrapper: { getByTestId },
      } = setup({ focused: false, onFocus: mockOnFocus });

      fireEvent.press(getByTestId('subheader'));

      expect(mockOnFocus).toHaveBeenCalled();
    });

    it("calls onFocus when pressing on Subheader's icon", () => {
      const mockOnFocus = jest.fn();
      const {
        wrapper: { getByTestId },
      } = setup({ focused: false, onFocus: mockOnFocus, icon: <View /> });

      fireEvent.press(getByTestId('icon'));

      expect(mockOnFocus).toHaveBeenCalled();
    });
  });
});


example/App.js
import React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';
import { createDrawerNavigator } from '@react-navigation/drawer';
import { DefaultTheme } from 'react-native-paper';
import ShopElectronicsScreen from './screens/ShopElectronics';

const Stack = createStackNavigator();
const Drawer = createDrawerNavigator();

const NavigationStack = () => {
  return (
    <Stack.Navigator
      initialRouteName="Home"
      screenOptions={{ gestureEnabled: false }}
    >
      <Stack.Screen
        name="ShopElectronics"
        component={ShopElectronicsScreen}
        options={{
          title: 'Shop Electronics',
          headerStyle: {
            backgroundColor: DefaultTheme.colors.primary,
            borderBottomWidth: 0,
            elevation: 0,
          },
          headerTintColor: '#FFF',
        }}
      />
    </Stack.Navigator>
  );
};

const App = () => {
  return (
    <NavigationContainer>
      <Drawer.Navigator>
        <Drawer.Screen name="Home" component={NavigationStack} />
      </Drawer.Navigator>
    </NavigationContainer>
  );
};

export default App;


src/Backdrop/Subheader.js
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import Touchable from 'react-native-platform-touchable';
import PropTypes from 'prop-types';

const Subheader = ({
  disabled,
  onPress,
  icon,
  title,
  titleStyle,
  TitleComponent,
}) => {
  return (
    <Touchable
      disabled={disabled}
      onPress={onPress}
      background={Touchable.Ripple('rgba(0, 0, 0, 0.2)', true)}
      accessibilityState={{ expanded: disabled }}
      accessibilityLabel={title}
      accessibilityRole="button"
    >
      <View style={styles.headerContainer}>
        {TitleComponent ? (
          <TitleComponent />
        ) : (
          <Text style={[styles.title, titleStyle]}>{title}</Text>
        )}
        {icon && (
          <Touchable
            onPress={onPress}
            centered
            style={styles.icon}
            background={Touchable.Ripple('rgba(0, 0, 0, 0.2)', true)}
            testID="icon"
            hitSlop={{ top: 20, left: 20, bottom: 20, right: 20 }}
            accessibilityRole="button"
          >
            {icon}
          </Touchable>
        )}
      </View>
    </Touchable>
  );
};

Subheader.propTypes = {
  disabled: PropTypes.bool,
  onPress: PropTypes.func.isRequired,
  icon: PropTypes.element,
  title: PropTypes.string.isRequired,
  TitleComponent: PropTypes.elementType,
  titleStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
};

Subheader.defaultProps = {
  disabled: false,
  icon: null,
  titleStyle: {},
  TitleComponent: null,
};

const styles = StyleSheet.create({
  headerContainer: {
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: 'rgba(0, 0, 0, 0.12)',
    marginHorizontal: 16,
    paddingVertical: 12,
  },
  title: {
    fontSize: 16,
    lineHeight: 24,
    color: 'rgba(0, 0, 0, 0.87)',
  },
  icon: {
    position: 'absolute',
    top: 12,
    right: 16,
    width: 24,
    height: 24,
  },
});

export default Subheader;


example/screens/ShopElectronics.js
import React, { useState } from 'react';
import { ScrollView, StyleSheet, View } from 'react-native';
import { DefaultTheme, List } from 'react-native-paper';
import { MaterialIcons } from '@expo/vector-icons';
import Backdrop from '../../src/Backdrop';

const ShopElectronicsScreen = () => {
  const [focusedBackdrop, setFocusedBackdrop] = useState(true);

  return (
    <View style={styles.container}>
      <ScrollView>
        <List.Item
          title="TV & Home Theaters"
          left={(props) => (
            <List.Icon {...props} color="#FFF" icon="television" />
          )}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
        <List.Item
          title="Computers"
          left={(props) => <List.Icon {...props} color="#FFF" icon="laptop" />}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
        <List.Item
          title="Cameras and Camcorders"
          left={(props) => <List.Icon {...props} color="#FFF" icon="camera" />}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
        <List.Item
          title="Mobile Phones"
          left={(props) => (
            <List.Icon {...props} color="#FFF" icon="cellphone-android" />
          )}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
        <List.Item
          title="Speakers"
          left={(props) => <List.Icon {...props} color="#FFF" icon="speaker" />}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
        <List.Item
          title="Video Games"
          left={(props) => (
            <List.Icon {...props} color="#FFF" icon="google-controller" />
          )}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
        <List.Item
          title="Movies"
          left={(props) => (
            <List.Icon {...props} color="#FFF" icon="filmstrip" />
          )}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
        <List.Item
          title="Music"
          left={(props) => (
            <List.Icon {...props} color="#FFF" icon="music-note" />
          )}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
        <List.Item
          title="Wearables"
          left={(props) => <List.Icon {...props} color="#FFF" icon="watch" />}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
        <List.Item
          title="Connected Home"
          left={(props) => (
            <List.Icon {...props} color="#FFF" icon="google-home" />
          )}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
        <List.Item
          title="Toys"
          left={(props) => (
            <List.Icon {...props} color="#FFF" icon="pinwheel" />
          )}
          onPress={() => {}}
          titleStyle={styles.listTitle}
        />
      </ScrollView>
      <Backdrop
        focused={focusedBackdrop}
        onFocus={() => setFocusedBackdrop(!focusedBackdrop)}
        title="All Products"
        icon={
          focusedBackdrop ? (
            <MaterialIcons name="keyboard-arrow-down" size={24} color="#000" />
          ) : (
            <MaterialIcons name="keyboard-arrow-up" size={24} color="#000" />
          )
        }
        backdropStyle={styles.backdropStyle}
      >
        <ScrollView>
          <List.Item title="Product 1" />
          <List.Item title="Product 2" />
          <List.Item title="Product 3" />
          <List.Item title="Product 4" />
          <List.Item title="Product 5" />
          <List.Item title="Product 6" />
          <List.Item title="Product 7" />
          <List.Item title="Product 8" />
          <List.Item title="Product 9" />
          <List.Item title="Product 10" />
          <List.Item title="Product 11" />
          <List.Item title="Product 12" />
          <List.Item title="Product 13" />
          <List.Item title="Product 14" />
        </ScrollView>
      </Backdrop>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: DefaultTheme.colors.primary,
  },
  listTitle: {
    color: '#FFF',
  },
  backdropStyle: {
    backgroundColor: '#F4F4F5',
  },
});

export default ShopElectronicsScreen;


src/Backdrop/Backdrop.js
import React, { useState, useEffect } from 'react';
import {
  LayoutAnimation,
  Platform,
  StyleSheet,
  UIManager,
  View,
} from 'react-native';
import PropTypes from 'prop-types';
import Subheader from './Subheader';

const Backdrop = ({
  children,
  focused,
  onFocus,
  title,
  icon,
  preset,
  titleStyle,
  backdropStyle,
}) => {
  const [flex, setFlex] = useState(focused ? 1 : 0);

  useEffect(() => {
    if (
      Platform.OS === 'android' &&
      UIManager.setLayoutAnimationEnabledExperimental
    ) {
      UIManager.setLayoutAnimationEnabledExperimental(true);
    }
  }, []);

  useEffect(() => {
    LayoutAnimation.configureNext(preset);

    if (focused) {
      setFlex(1);
    } else {
      setFlex(0);
    }
  }, [focused, preset]);

  return (
    <View focusable={focused} pointerEvents="box-none" style={styles.overlay}>
      <View
        style={[styles.backdrop, { flex }, backdropStyle]}
        testID="backdrop"
      >
        <Subheader
          disabled={focused}
          onPress={onFocus}
          icon={icon}
          title={title}
          titleStyle={titleStyle}
          testID="subheader"
        />
        <View
          style={{
            flex: 1,
          }}
        >
          <View style={styles.children}>{children}</View>
        </View>
      </View>
    </View>
  );
};

Backdrop.propTypes = {
  children: PropTypes.node,
  focused: PropTypes.bool,
  onFocus: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  icon: PropTypes.element,
  preset: PropTypes.object,
  titleStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  backdropStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
};

Backdrop.defaultProps = {
  children: null,
  focused: true,
  icon: null,
  preset: LayoutAnimation.Presets.easeInEaseOut,
  titleStyle: {},
  backdropStyle: {},
};

const styles = StyleSheet.create({
  overlay: {
    ...StyleSheet.absoluteFill,
    justifyContent: 'flex-end',
  },
  backdrop: {
    backgroundColor: '#FFF',
    borderTopLeftRadius: 16,
    borderTopRightRadius: 16,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.2,
    shadowRadius: 1.41,
    elevation: 2,
  },
  children: {
    marginHorizontal: 16,
    flex: 1,
  },
});

export default Backdrop;
